cmd/fanoutsidecar/main.go

```diff
@@ -125,57 +127,87 @@ func main() {
 	}
 }
 
-func setupConfigMapNoticer(logger *zap.Logger, configUpdated swappable.UpdateConfig) (manager.Manager, error) {
+func setupChannelWatcher(logger *zap.Logger, configUpdated swappable.UpdateConfig) (manager.Manager, error) {
 	mgr, err := manager.New(config.GetConfigOrDie(), manager.Options{})
 	if err != nil {
-		logger.Error("Error starting manager.", zap.Error(err))
+		logger.Error("Error creating new maanger.", zap.Error(err))
 		return nil, err
 	}
-
-	switch configMapNoticer {
-	case cmnfVolume:
-		err = setupConfigMapVolume(logger, mgr, configUpdated)
-	case cmnfWatcher:
-		err = setupConfigMapWatcher(logger, mgr, configUpdated)
-	default:
-		err = fmt.Errorf("need to provide the --config_map_noticer flag (valid values are %s)", configMapNoticerValues())
-	}
-	if err != nil {
+	if err = v1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
+		logger.Error("Error while adding eventing scheme to manager.", zap.Error(err))
 		return nil, err
 	}
+	channelwatcher.New(mgr, logger, updateChannelConfig(configUpdated))
 
 	return mgr, nil
 }
 
-func setupConfigMapVolume(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	cmn, err := filesystem.NewConfigMapWatcher(logger, filesystem.ConfigDir, configUpdated)
-	if err != nil {
-		logger.Error("Unable to create filesystem.ConifgMapWatcher", zap.Error(err))
-		return err
-	}
-	if err = mgr.Add(cmn); err != nil {
-		logger.Error("Unable to add the config map watcher", zap.Error(err))
-		return err
+func updateChannelConfig(updateConfig swappable.UpdateConfig) channelwatcher.WatchHandlerFunc {
+	return func(ctx context.Context, c client.Client, chanNamespacedName types.NamespacedName) error {
+		channels, err := listAllChannels(ctx, c)
+		if err != nil {
+			logging.FromContext(ctx).Info("Unable to list channels", zap.Error(err))
+			return err
+		}
+		config := multiChannelFanoutConfig(channels)
+		return updateConfig(config)
 	}
-	return nil
 }
 
-func setupConfigMapWatcher(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	kc, err := kubernetes.NewForConfig(mgr.GetConfig())
-	if err != nil {
-		return err
+func listAllChannels(ctx context.Context, c client.Client) ([]v1alpha1.Channel, error) {
+	channels := make([]v1alpha1.Channel, 0)
+	for {
```

Comment:
It might be worth a comment here indicating that this is a do... while loop on `opts.Raw.Continue`.

Reply:
This is probably copied from other code that added pagination support defensively (@Harwayne may be able to provide context). Based on my reading of [the original design for apiserver pagination](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-chunking.md), it appears @evankanderson is correct: pagination is opt-in by specifying the `limit` parameter.

Reply:
I am paranoid. I don't think it gets used today. I believe @akashrv looked into the Controller Runtime code in particular and confirmed that it isn't used. But the fact that it is in the interface makes me wary of leaving it out. I agree that ListOptions.Limit clearly talks about Continue with regards to having a limit set, but ListOptions.Continue doesn't specify that is the _only_ time Continue will be used.

I agree with your reading of the origin design that this is purely an opt-in feature and that not providing a limit means get everything (even if that is 500 MB). Having been shown this, I am OK removing it.

Reply:
Sounds good. I'll simplify it and revert it to how it was in first iteration.